src/runtime-rs/crates/resource/src/network/endpoint/physical_endpoint.rs

```diff
@@ -0,0 +1,140 @@
+
+        // get vendor and device id from pci space (sys/bus/pci/devices/$bdf)
+        let iface_device_path = sys_pci_devices_path.join(&bdf).join("device");
+        let contents = std::fs::read(&iface_device_path)
+            .context(format!("read device path {:?}", &iface_device_path))?;
+        let device_id = String::from_utf8(contents).context("device id from utf8")?;
+
+        let iface_vendor_path = sys_pci_devices_path.join(&bdf).join("vendor");
+        let contents = std::fs::read(&iface_vendor_path)
+            .context(format!("read vendor path {:?}", &iface_vendor_path))?;
+        let vendor_id = String::from_utf8(contents).context("vendor id from utf8")?;
+
+        Ok(Self {
+            iface_name: name.to_string(),
+            hard_addr: utils::get_mac_addr(hardware_addr),
+            vendor_device_id: VendorDevice::new(&vendor_id, &device_id),
+            driver,
+            bdf,
+        })
+    }
+}
+
+#[async_trait]
+impl Endpoint for PhysicalEndpoint {
+    async fn name(&self) -> String {
+        self.iface_name.clone()
+    }
+
+    async fn hardware_addr(&self) -> String {
+        self.hard_addr.clone()
+    }
+
+    async fn attach(&self, hypervisor: &dyn Hypervisor) -> Result<()> {
+        // bind physical interface from host driver and bind to vfio
+        device::bind_device_to_vfio(
+            &self.bdf,
+            &self.driver,
+            &self.vendor_device_id.vendor_device_id(),
+        )
+        .context(format!(
+            "bind physical endpoint from {} to vfio",
+            &self.driver
+        ))?;
+
+        // set vfio's bus type, pci or mmio. Mostly use pci by default.
+        let mode = match self.driver.as_str() {
+            "virtio-pci" => "mmio",
+            _ => "pci",
+        };
+
+        // add vfio device
+        let d = device::Device::Vfio(device::VfioConfig {
+            id: format!("physical_nic_{}", self.name().await),
+            sysfs_path: "".to_string(),
+            bus_slot_func: self.bdf.clone(),
+            mode: device::VfioBusMode::new(mode)
+                .context(format!("new vfio bus mode {:?}", mode))?,
+        });
+        hypervisor.add_device(d).await.context("add device")?;
+        Ok(())
+    }
+
+    // Detach for physical endpoint unbinds the physical network interface from vfio-pci
+    // and binds it back to the saved host driver.
+    async fn detach(&self, _hypervisor: &dyn Hypervisor) -> Result<()> {
+        // bind back the physical network interface to host.
+        // we need to do this even if a new network namespace has not
+        // been created by virt-containers.
+
+        // we do not need to enter the network namespace to bind back the
+        // physical interface to host driver.
+        device::bind_device_to_host(
+            &self.bdf,
+            &self.driver,
+            &self.vendor_device_id.vendor_device_id(),
+        )
+        .context(format!(
```

Comment:
with_context same reason with above

Reply:
Thanks for your review, since the `context` has been used everywhere, I'll raise an issue(#4129) to change this to `with_context`.